app/scripts/controllers/network/provider-api-tests/shared-tests.js

```diff
@@ -750,100 +2176,144 @@ export function testsForRpcMethodSupportingBlockParam(
+              request,
+              blockParamIndex,
+              '0x100',
+            ),
+            error: errorMessage,
+            times: 5,
+          });
+
+          const promiseForResult = withClient(
+            { providerType },
+            async ({ makeRpcCall, clock }) => {
+              return await waitForPromiseToBeFulfilledAfterRunningAllTimers(
+                makeRpcCall(request),
+                clock,
+              );
+            },
+          );
+
+          await expect(promiseForResult).rejects.toThrow(
+            buildJsonRpcEngineEmptyResponseErrorMessage(method),
+          );
         });
+      });
+    }
+
+    if (providerType === 'infura') {
+      it('retries the request to the RPC endpoint up to 5 times if an "ECONNRESET" error is thrown while making the request, returning the successful result if there is one on the 5th try', async () => {
+        await withMockedCommunications({ providerType }, async (comms) => {
+          const request = { method };
+
+          // The first time a block-cacheable request is made, the
+          // block-cache middleware will request the latest block number
+          // through the block tracker to determine the cache key. Later,
+          // the block-ref middleware will request the latest block number
+          // again to resolve the value of "latest", but the block number is
+          // cached once made, so we only need to mock the request once.
+          comms.mockNextBlockTrackerRequest({ blockNumber: '0x100' });
+          // The block-ref middleware will make the request as specified
+          // except that the block param is replaced with the latest block
+          // number.
+          //
+          // Here we have the request fail for the first 4 tries, then
+          // succeed on the 5th try.
+          comms.mockRpcCall({
+            request: buildRequestWithReplacedBlockParam(
+              request,
+              blockParamIndex,
+              '0x100',
+            ),
+            error: 'ECONNRESET: Some message',
+            times: 4,
+          });
+          comms.mockRpcCall({
+            request: buildRequestWithReplacedBlockParam(
+              request,
+              blockParamIndex,
+              '0x100',
+            ),
+            response: {
+              result: 'the result',
+              httpStatus: 200,
+            },
+          });
 
-        expect(results).toStrictEqual([
-          mockResults[0],
-          mockResults[0],
-          mockResults[0],
-        ]);
+          const result = await withClient(
+            { providerType },
+            async ({ makeRpcCall, clock }) => {
+              return await waitForPromiseToBeFulfilledAfterRunningAllTimers(
+                makeRpcCall(request),
+                clock,
+              );
+            },
+          );
+
+          expect(result).toStrictEqual('the result');
+        });
       });
-    });
 
-    if (blockParamType === 'none') {
```

Comment:
I noticed this `if` statement while perusing this file. I think I left this `if` statement in a previous PR for debugging purposes, but this should not be here — or rather, this code should not be wrapped in an `if` statement. I've unwrapped the code in this PR, but have also extracted it to another PR: https://github.com/MetaMask/metamask-extension/pull/16906